OpenCL_C.txt

```diff
@@ -10052,8 +10052,8 @@ NOTE: The functionality described in this section <<unified-spec, requires>>
 support for OpenCL C 2.0, or OpenCL C 3.0 or newer and the {opencl_c_pipes} feature.
 
 A pipe is identified by specifying the `pipe` keyword with a type.
-The data type specifies the size of each packet in the pipe.
-The `pipe` keyword is a type modifier.
+The data type specifies the type of each element in the pipe.
```

Comment:
I'm wondering whether we want to keep something along the lines of "specifies the type, **and thus size**" or if it's "obvious".

Reply:
I find it implicit since every type has its size.

Reply:
Yeah, it does seem implicit. Does it need to be explicit, though? The API spec talks about packet sizes, and I don't see anything saying that the read and write ends have to use identical data types, but I would expect the packet size would need to match.

The concern would be that someone might assume it's valid to write int4 values to one end and read scalar ints from the other, but as far as I can see, this would be invalid.

Reply:
The OpenCL C specification consistently uses the term "packet" when referring to pipes. It may be worth discussing whether "element" would be clearer, it would make sense to use a single term consistently. Perhaps the term "element" was originally avoided to prevent confusion between "pipe elements" and "vector elements".

Reply:
FWIW, the OpenCL API spec also refers almost exclusively to pipe "packets" and not pipe "elements".

The OpenCL C spec seems to hint that they are equivalent:

> When it is applied to another type T, the result is a pipe type whose elements (or packets) are of type T.

We're getting a bit out in the weeds here from the purpose of this PR, but I also wonder how the `pipe_packet_size` specified when the pipe is created relates to the OpenCL C pipe type.  Does the size have to match?  Or, it OK if the pipe packet size is bigger than the OpenCL C pipe type?  What happens if the pipe packet size is smaller than the OpenCL C pipe type?

Reply:
As for OpenCL C I would prefer to use `element` as it is more common concept in C-based lnguages.

>We're getting a bit out in the weeds here from the purpose of this PR, but I also wonder how the pipe_packet_size specified when the pipe is created relates to the OpenCL C pipe type. Does the size have to match? Or, it OK if the pipe packet size is bigger than the OpenCL C pipe type? What happens if the pipe packet size is smaller than the OpenCL C pipe type?

My understanding is that pipe type itself is opaque and its size is implementation defined and allowed to be different from its element/packet type. In clang it is just a pointer type to a struct that comntains implementation of pipe data structure.

Reply:
I can see thare are a lot more we can clarify for pipes - should we create separate issues/PRs?